crates/oxc_linter/src/rules/no_dupe_class_members.rs

```diff
@@ -0,0 +1,195 @@
+use oxc_ast::{
+    ast::{ClassElement, MethodDefinitionKind},
+    AstKind, Atom, GetSpan, Span,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use rustc_hash::FxHashMap;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-dupe-class-members): Duplicate class member: {0:?}")]
+#[diagnostic(
+    severity(warning),
+    help(
+        "The last declaration overwrites previous ones, remove one of them or rename if both should be retained"
+    )
+)]
+struct NoDupeClassMembersDiagnostic(
+    Atom, /*Class member name */
+    #[label("{0:?} is previously declared here")] pub Span,
+    #[label("{0:?} is re-declared here")] pub Span,
+);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoDupeClassMembers;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow duplicate class members
+    ///
+    /// ### Why is this bad?
+    /// If there are declarations of the same name in class members,
+    /// the last declaration overwrites other declarations silently. It can cause unexpected behaviors.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// class A {
+    ///   foo() { console.log("foo") }
+    ///   foo = 123;
+    /// }
+    /// let a = new A();
+    /// a.foo() // Uncaught TypeError: a.foo is not a function
+    /// ```
+    NoDupeClassMembers,
+    correctness
+);
+
+impl Rule for NoDupeClassMembers {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::Class(class) = node.get().kind() else { return; };
+
+        let mut property_table = PropertyTable::default();
```

Comment:
This is the right way, because we need to check all the classes anyway.

One small improvement is to reserve capacity by the total number of class members.